controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesInt96Counter.java

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.controller.store.stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.lang.AtomicInt96;
+import io.pravega.common.lang.Int96;
+import io.pravega.common.tracing.TagLogger;
+import io.pravega.controller.store.PravegaTablesStoreHelper;
+import io.pravega.controller.store.VersionedMetadata;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import static io.pravega.shared.NameUtils.TRANSACTION_ID_COUNTER_TABLE;
+
+/**
+ * A Pravega tables based int 96 counter. At bootstrap, and upon each refresh, it retrieves a unique starting counter from the
+ * Int96 space. It does this by getting and updating an Int96 value stored in Pravega Table. It updates the table value
+ * with the next highest value that others could access.
+ * So upon each retrieval from Pravega table, it has a starting counter value and a ceiling on counters it can assign until a refresh
+ * is triggered. Any caller requesting for nextCounter will get a unique value from the range owned by this counter.
+ * Once it exhausts the range, it refreshes the range by contacting Pravega Tables and repeating the steps described above.
+ */
+
+public class PravegaTablesInt96Counter implements Int96Counter {
```

Comment:
How much common logic from this class can be moved to `Int96Counter` and shared across the ZK and KVTables implementation?

Reply:
I have moved common code in abstract class.